curp/src/client_new/tests.rs

```diff
@@ -0,0 +1,645 @@
+                        leader_id: Some(3), // imagine this node is a old leader
+                        term: 1,            // with the old term
+                        cluster_id: 123,
+                        members: vec![
+                            Member::new(0, "S0", vec!["B0".to_owned()], false),
+                            Member::new(1, "S1", vec!["B1".to_owned()], false),
+                            Member::new(2, "S2", vec!["B2".to_owned()], false),
+                            Member::new(3, "S3", vec!["B3".to_owned()], false),
+                            Member::new(4, "S4", vec!["B4".to_owned()], false),
+                        ],
+                        cluster_version: 1,
+                    },
+                    4 => FetchClusterResponse {
+                        leader_id: Some(3), // imagine this node is a old follower of old leader(3)
+                        term: 1,            // with the old term
+                        cluster_id: 123,
+                        members: vec![],
+                        cluster_version: 1,
+                    },
+                    _ => unreachable!("there are only 5 nodes"),
+                };
+                Ok(tonic::Response::new(resp))
+            });
+    });
+    let unary = Unary::<TestCommand>::new(connects, None, None);
+    assert!(unary.local_connect().is_none());
+    let res = unary.fetch_cluster(true).await.unwrap_err();
+    // only server(0, 1)'s responses are valid, less than majority quorum(3), got a mocked RpcTransport to retry
+    assert_eq!(res, CurpError::RpcTransport(()));
+}
+
+#[traced_test]
+#[tokio::test]
+async fn test_unary_fast_round_works() {
+    let connects = init_mocked_connects(5, |id, conn| {
+        conn.expect_propose().return_once(move |_req, _timeout| {
+            let resp = match id {
+                0 => ProposeResponse::new_result::<TestCommand>(&Ok(TestCommandResult::default())),
+                1 | 2 | 3 => ProposeResponse::new_empty(),
+                4 => return Err(CurpError::key_conflict()),
+                _ => unreachable!("there are only 5 nodes"),
+            };
+            Ok(tonic::Response::new(resp))
+        });
+    });
+    let unary = Unary::<TestCommand>::new(connects, None, None);
+    let res = unary
+        .fast_round(ProposeId(0, 0), &TestCommand::default())
+        .await
+        .unwrap()
+        .unwrap();
+    assert_eq!(res, TestCommandResult::default());
+}
+
+#[traced_test]
+#[tokio::test]
+async fn test_unary_fast_round_return_early_err() {
+    for early_err in [
+        CurpError::duplicated(),
+        CurpError::shutting_down(),
+        CurpError::invalid_config(),
+        CurpError::node_already_exists(),
+        CurpError::node_not_exist(),
+        CurpError::learner_not_catch_up(),
+        CurpError::expired_client_id(),
+        CurpError::wrong_cluster_version(),
+        CurpError::redirect(Some(1), 0),
+    ] {
+        assert!(early_err.return_early());
```

Comment:
The `CurpError` is non-exhaustive, and one may forget to update these tests after updating the variants. Maybe we need a way to explicitly define these errors?

Reply:
Maybe we could defined a `fn values() -> Vec<CurpError>` method in `CurpError` ?

Reply:
I think adding a note to the definition of `CurpError` will also do the trick.

Reply:
added